x/bank/keeper/keeper.go

```diff
@@ -145,13 +145,17 @@ func (k BaseKeeper) DelegateCoins(ctx sdk.Context, delegatorAddr, moduleAccAddr
 	}
 
 	balances := sdk.NewCoins()
+	// Funds in a vesting account can still be delegated. Funds locked by other means cannot. Or if they should
+	// be available, the author of that GetLockedCoinsFn can check the context against HasVestingLockedBypass too.
+	locked := k.LockedCoins(types.WithVestingLockedBypass(ctx), delegatorAddr)
```

Comment:
Did you consider changing this flow such that ‘unvested’ coin amounts are added to the available balance for delegations instead of the exceptional cases around locks?

Reply:
By using `LockedCoins` with a bypass, it skips the `UnvestedCoins` lookup altogether. If I were to start with spendable and add those back in, that lookup would end up happening twice. Additionally, there isn't a single-denom version of the `SpendableCoins` function, so using it can cause lookup of data that's not needed.

I thought about creating single-denom versions of "spendable", "locked", and "unvested", but held off because of the bank keeper API changes needed. As it is now, the injected locked-coins functions might be doing lookups that aren't actually needed, but I figured we could do some refactoring around that later if we think it's going to be a problem.

Reply:
I was thinking of perhaps a more invasive change to the vesting account... essentially having the vesting account calculate and update a "hold amount" that is locked (just like everything else) ... but the "delegation" component using an available balance that was based on the current available + the amount of any "vesting account holds".  
Along with this the delegation module would of course create a hold for any delegations made as well...

This setup would result in an overlapping hold on specific balances though.

Reply:
Above might be a future refinement if the concept makes sense... because in my mind the vesting accounts are horribly designed and a serious hack on things which the hold module make large strides towards fixing.